File: core/support/src/main/java/com/reservation/support/enums/SocialLoginProvider.java
package com.reservation.support.enums;

public enum SocialLoginProvider {
	GOOGLE, GITHUB;

	private static final String FALLBACK_BASE_URL = "https://hotel-reservation-frontend.com/social-fail";
	private static final String REDIRECT_URL = "https://hotel-reservation-frontend.com/login/success";
	private static final String SOCIAL_SIGNUP_URL = "https://hotel-reservation-frontend.com/signup?email=";

	public String getRedirectUrl() {
		return REDIRECT_URL;
	}

	public String getSocialSignupUrl(String email) {
		return SOCIAL_SIGNUP_URL + email;
	}

	public String fallbackRedirectUrl(String reason) {
		String queryString = reason != null ? "?reason=" + reason : "";
		return FALLBACK_BASE_URL + queryString;
	}

	public String fallbackUnknown() {
		return fallbackRedirectUrl("UNKNOWN");
	}
}

File: core/support/src/main/kotlin/com/reservation/support/enums/SocialLoginProvider.kt
package com.reservation.support.enums

enum class SocialLoginProvider {
    GOOGLE, GITHUB;

    val redirectUrl: String
        get() = REDIRECT_URL

    fun getSocialSignupUrl(email: String): String {
        return SOCIAL_SIGNUP_URL + email
    }

    fun fallbackRedirectUrl(reason: String?): String {
        val queryString = if (reason != null) "?reason=$reason" else ""
        return FALLBACK_BASE_URL + queryString
    }

    fun fallbackUnknown(): String {
        return fallbackRedirectUrl("UNKNOWN")
    }

    companion object {
        private const val FALLBACK_BASE_URL: String = "https://hotel-reservation-frontend.com/social-fail"
        private const val REDIRECT_URL: String = "https://hotel-reservation-frontend.com/login/success"
        private const val SOCIAL_SIGNUP_URL: String = "https://hotel-reservation-frontend.com/signup?email="
    }
}
